gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java

```diff
@@ -132,16 +146,70 @@ public static void waitJobCompletion(
     while (!timeoutInSeconds.isPresent() || System.currentTimeMillis() <= endTime) {
       WorkflowContext workflowContext = TaskDriver.getWorkflowContext(helixManager, workFlowName);
       if (workflowContext != null) {
-        org.apache.helix.task.TaskState helixJobState = workflowContext.getJobState(TaskUtil.getNamespacedJobName(workFlowName, jobName));
-        if (helixJobState == org.apache.helix.task.TaskState.COMPLETED ||
-            helixJobState == org.apache.helix.task.TaskState.FAILED ||
-            helixJobState == org.apache.helix.task.TaskState.STOPPED) {
+        log.info("workflowContext is {}", workflowContext.getJobState(TaskUtil.getNamespacedJobName(workFlowName, jobName)));
+        switch (workflowContext.getJobState(TaskUtil.getNamespacedJobName(workFlowName, jobName))) {
+          case STOPPED:
+            // user requested cancellation, which is executed by executeCancellation()
+            log.info("Job {} is cancelled, it will be deleted now.", jobName);
+            new TaskDriver(helixManager).deleteAndWaitForCompletion(workFlowName, 10000L);
+            return;
+          case FAILED:
+          case COMPLETED:
           return;
         }
+      } else {
+        // We have waited for WorkflowContext to get initialized,
+        // so it is found null here, it must have been deleted in job cancellation process.
+        log.info("WorkflowContext not found. Job is probably cancelled.");
+        return;
       }
+      log.info("Waiting for job {} to complete...", jobName);
       Thread.sleep(1000);
     }
 
     throw new TimeoutException("task driver wait time [" + timeoutInSeconds + " sec] is expired.");
   }
-}
+
+  public static void handleJobTimeout(String workFlowName, String jobName, HelixManager helixManager,
+      Either<GobblinHelixJobLauncher, GobblinHelixDistributeJobExecutionLauncher> jobLauncher,
```

Comment:
It may be cleaner to create a specific wrapper object like GobblinHelixJobLauncherHolder to hold the two object types or create an interface with the cancel method that both implement.